cmake/ImportBSON.cmake

```diff
@@ -0,0 +1,170 @@
+   # Disable AWS_AUTH, to prevent it from building the kms-message symbols, which we build ourselves
+   set (ENABLE_MONGODB_AWS_AUTH OFF CACHE BOOL "Disable kms-message content in mongoc for libmongocrypt" FORCE)
+   # Disable install() for the libbson static library. We'll do it ourselves
+   set (ENABLE_STATIC BUILD_ONLY)
+   # Disable over-alignment of bson types
+   set (ENABLE_EXTRA_ALIGNMENT ${_extra_alignment_default} CACHE BOOL "Toggle extra alignment of bson_t")
+   # Add the subdirectory as a project. EXCLUDE_FROM_ALL to inhibit building and installing of components unless requested
+   add_subdirectory ("${MONGOCRYPT_MONGOC_DIR}" _mongo-c-driver EXCLUDE_FROM_ALL)
+endfunction ()
+
+# Do the add_subdirectory() in a function to isolate variable scope
+_import_bson_add_subdir ()
+
+# Define interface targets to be used to control the libbson used at both build and import time.
+# Refer to mongocrypt-config.cmake to see how these targets are used by consumers
+add_library (_mongocrypt-libbson_for_static INTERFACE)
+add_library (_mongocrypt-libbson_for_shared INTERFACE)
+add_library (_mongocrypt::libbson_for_static ALIAS _mongocrypt-libbson_for_static)
+add_library (_mongocrypt::libbson_for_shared ALIAS _mongocrypt-libbson_for_shared)
+install (
+   TARGETS _mongocrypt-libbson_for_static _mongocrypt-libbson_for_shared
+   EXPORT mongocrypt_targets
+)
+
+# Link to the requested libbson, only exporting that usage for the local build tree.
+# The mongocrypt-config file will later add the appropriate link library for downstream
+# users during find_package()
+if (USE_SHARED_LIBBSON)
+   target_link_libraries (_mongocrypt-libbson_for_shared INTERFACE $<BUILD_INTERFACE:bson_shared>)
+else ()
+   target_link_libraries (_mongocrypt-libbson_for_shared INTERFACE $<BUILD_INTERFACE:bson_static>)
+endif ()
+# libbson_for_static always links to the static libbson:
+target_link_libraries (_mongocrypt-libbson_for_static INTERFACE $<BUILD_INTERFACE:bson_static>)
+
+# And an alias to the mongoc target for use in some test cases
+add_library (_mongocrypt::mongoc ALIAS mongoc_static)
+# Workaround: Embedded mongoc_static does not set its INCLUDE_DIRECTORIES for user targets
+target_include_directories (mongoc_static
+   PUBLIC
+      "$<BUILD_INTERFACE:${MONGOCRYPT_MONGOC_DIR}/src/libmongoc/src>"
+      "$<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/_mongo-c-driver/src/libmongoc/src/mongoc>"
+   )
+
+# Put the libbson dynamic library into the current binary directory (plus possible config suffix).
+# This ensures that libbson DLL will resolve on Windows when it searches during tests
+set_property(TARGET bson_shared PROPERTY RUNTIME_OUTPUT_DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}")
+
+if (ENABLE_STATIC)
+   # We are going to build a static libmongocrypt.
+   # We want the static libbson target from the embedded mongoc. Enable the static library as
+   # part of "all", and install the archive alongside the rest of our static libraries.
```

Comment:
For reference, up until now we have only included the static libbson, as `libbson-static-1.0.a`, in the Debian packages and RPMs that we build for publication in our PPA.  The specific reason for this is that we do not publish libmongoc/libbson packages via PPA and so we must account for the very likely scenario where a libmongocrypt user is installing from the PPA to a platform that doesn't have a recent libbson or doesn't have it at all.

Making the libmongocrypt build install a `libbson-static-for-libmongocrypt.a` feels like not the right thing.  I'd rather see us go the plain vanilla external project route.  If the user does not provide an adequate libbson (either in system directories or in some other user-specified location), then we download the C driver, build the necessary components and install from the C driver build to the installation target specified by the user for the top-level libmongocrypt build.  We can then document for the user that if a suitable libbson isn't provided, that one will be fetched by the libmongocrypt build and will be installed alongside the libmongocrypt components.

Reply:
The `libbson-static-for-libmongocrypt` file is to replace the `libbson-static` that was being copied and installed manually by the build shell scripts (The one created for CDRIVER-3187). The reason it is given the specific name here is because the Debian package build (correctly) errors because there's already a package (the actual libbson-dev package) that installs a file at the exact same filepath.

As for how to handle linking `mongocrypt_static` properly when libbson is not easily available, that's a good question. Previously the installed libmongocrypt would just rely on the `find_package()` result to propagate correctly, which isn't reliable in general:

- If the `find_package([some-libbson])` in `libmongocrypt` generated import targets (i.e. `mongo::bson[_static]`), the exported libmongocrypt targets would have link libraries of `mongo::bson[_static]`, and it was up to the user to ensure that those were visible by the time their CMake project finished configuring (i.e. the user had to `find_package()` to get libbson).
- If the `find_package()` generated the old `${BSON_FOO}` variables, then the generated libmongocrypt targets would have each of the libbson attributes repeated as interface properties and would have an explicit link to whatever the found libbson library file was found. I don't think this codepath is very common since most users are probably on a new enough libbson that the `mongo::bson` targets are installed.

In this changeset, mongocrypt and mongocrypt_static link-only to targets `_mongocrypt-libbson_for_shared` and `_mongocrypt-libbson_for_static`, respectively. These are interface targets that are constructed such that their installed version exports no usage requirements in the `mongocrypt-targets.cmake` files. Instead, `mongocrypt-config.cmake` does additional work to ensure that these interface targets point to the proper libbson files so that the link will execute successfully.

Of course, as in your attempt, this fails to work, because `mongocrypt-config.cmake` cannot find a `libbson` static library to attach to `_mongocrypt-libbson_for_static`. Do you have a `libbson-static-1.0.a` library installed? If so, the `find_library()` call should have succeeded. If the `find_library()` call fails, we _could_ issue a warning message to users, but it may be irrelevant for some users that do not wish to link the libmongocrypt statically.

Handling this properly is difficult when you attempt to build and install both static and dynamic libraries as a single package and cannot assume the way that the user wishes to link their final executable.  For example:

1. If a user wants to statically link with libmongocrypt, then:
   1. Static libmongocrypt _always_ requires libbson definitions to be provided when CMake generates the runtime binaries.
   2. If the user wishes to link libbson statically, then `mongo::mongocrypt_static` must link_libraries to a static libbson.
   3. If the user wishes to link libbson dynamically, then `mongo::mongocrypt_static` must link_libraries to a dynamic libbson.
   4. If the user doesn't care about linking to libbson for their final application: Too bad: the static libmongocrypt _requires_ that the libbson definitions be provided. We can attempt to do this transparently for the user in `mongocrypt-config.cmake`, but should the default be to statically link to libbson? Or dynamically? In the current changeset, this defaults to try `mongo::bson_static`, or to fall back and link against a `libbson-static-1.0` via `find_library()`.
2. If the user wants to dynamically link libmongocrypt:
   1. If the libmongocrypt was built with `USE_SHARED_LIBBSON`, then `mongo::mongocrypt` requires that the final application link to a libbson dynamic library, as the `mongo::mongocrypt` binary contains an `SONAME` reference to libbson. Statically linking with libbson is not an option for the user in this case.
   2. If the libmongocrypt was build with `USE_SHARED_LIBBSON=OFF` (the default), then there are no requirements on transitively linking libbson, and the user can do as they wish.

Alternatively, in the case of `mongo::mongocrypt_static`, we could forego attempting to solve this entirely and have a documentation-only requirement that the user bring-their-own libbson symbols for their runtime binaries. This would generate link errors unless the user is linking to libbson in some way. This would be the simplest solution, actually, and now I'm leaning towards it.

Reply:
I would also lean toward the simpler documentation-only solution.  The user who is consuming libmongocrypt directly is most likely also consuming the C and/or C++ driver directly.  It seems unlikely that they would be trying to build only libmongocrypt without the need for any other driver support (not saying it would never happen, just that it seems unlikely to be common).  The other languages (Ruby, Node, etc.,) provide the appropriate shared and/or static libraries within their packages (which they obtain from the Evergreen build), so expecting that someone who directly consumes libmongocrypt to also provide their own libbson (static and/or dynamic as suits their particular use case) seems like a perfectly reasonable expectation, especially if in the common case they would already have to build it.

Reply:
> Do you have a libbson-static-1.0.a library installed?

I do not. 

> Alternatively, in the case of mongo::mongocrypt_static, we could forego attempting to solve this entirely and have a documentation-only requirement that the user bring-their-own libbson symbols for their runtime binaries. This would generate link errors unless the user is linking to libbson in some way. This would be the simplest solution, actually, and now I'm leaning towards it.

I am also in favor of the proposed documentation-only solution.